src/Features/Core/Portable/ExtractClass/AbstractExtractClassRefactoringProvider.cs

```diff
@@ -0,0 +1,105 @@
+
+        public AbstractExtractClassRefactoringProvider() : this(service: null)
+        {
+        }
+
+        public AbstractExtractClassRefactoringProvider(IExtractClassOptionsService? service)
+        {
+            _optionsService = service;
+        }
+
+        protected abstract Task<SyntaxNode?> GetSelectedNodeAsync(CodeRefactoringContext context);
+        protected abstract Task<SyntaxNode?> GetSelectedClassDeclarationAsync(CodeRefactoringContext context);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var optionsService = _optionsService ?? context.Document.Project.Solution.Workspace.Services.GetService<IExtractClassOptionsService>();
+            if (optionsService is null)
+            {
+                return;
+            }
+
+            // If we register the action on a class node, no need to find selected members. Just allow
+            // the action to be invoked with the dialog and no selected members
+            var action = await TryGetClassActionAsync(context, optionsService).ConfigureAwait(false)
+                ?? await RegisterMemberActionAsync(context, optionsService).ConfigureAwait(false);
+
+            if (action != null)
+                context.RegisterRefactoring(action, action.Span);
+        }
+
+        private async Task<ExtractClassWithDialogCodeAction?> RegisterMemberActionAsync(CodeRefactoringContext context, IExtractClassOptionsService optionsService)
+        {
+            var selectedMemberNode = await GetSelectedNodeAsync(context).ConfigureAwait(false);
+            if (selectedMemberNode is null)
+            {
+                return null;
+            }
+
+            var (document, span, cancellationToken) = context;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var selectedMember = semanticModel.GetDeclaredSymbol(selectedMemberNode);
+            if (selectedMember is null || selectedMember.ContainingType is null)
+            {
+                return null;
+            }
+
+            // Use same logic as pull members up for determining if a selected member
+            // is valid to be moved into a base
+            if (!MemberAndDestinationValidator.IsMemberValid(selectedMember))
+            {
+                return null;
+            }
+
+            var containingType = selectedMember.ContainingType;
+
+            // Can't extract to a new type if there's already a base. Maybe
+            // in the future we could inject a new type inbetween base and
+            // current
+            if (containingType.BaseType?.SpecialType != SpecialType.System_Object)
+            {
+                return null;
+            }
+
+            return new ExtractClassWithDialogCodeAction(document, span, optionsService, selectedMember.ContainingType, selectedMember);
```

Comment:
`selectedMember.ContainingType` seems redundant given that you pass selectedMember